### DISCUSSION TITLE
[dove] around hook

### DISCUSSION THREAD

**Discussion by anonimusprogramus** [feathersjs/feathers#2922] [Q&A]
No issue here, just need a clear understanding.

I thought `around hook` was **first & last** executed hook; turns out it wraps every before-after-error hook or service, so it will be executed more than once.

If I need an app-level `"setup & teardown"` that only **run once** upon user request (via socket) -- where is the proper place to put it, can it be inside a hook?

Thanks

**Comment by daffl** [comment]
`around` hooks will only be executed once per method call. This usually is the same as a user request (you can check where the request came from in `params.provider`). `setup` and `teardown` will only be executed once on server startup and shutdown.

**Comment by anonimusprogramus** [comment]
Please cmiiw, to my understanding, if an outlet page has

1. outlet's name & tag (will call outlet service)
2. outlet's product list (will call product service)
3. client's total items in cart (will call order service)

then a user request will execute `around hook` **three times** and then sending the response.

Is that right?

So if we want to run something before 1 and after 3, where can we do that?

_(It's like wrapping a user request, that's why I said "setup & teardown"-like :sweat_smile:)_

**Reply by daffl** [reply]
It sounds like you want a custom service that can do those multiple steps with one user request.

**Reply by anonimusprogramus** [reply]
Yes, you're right, I wrap those 3 service calls within another custom service :sweat_smile: 

And set a before hook for checking JWT. So total will be 5 times `around hook` calls.

I don't know if this (creating a custom method service only to do multiple calls) is a good or bad practice.

Anyway, it's running well. Thanks.
